python/lsst/ctrl/mpexec/singleQuantumExecutor.py

```diff
@@ -163,6 +182,24 @@ def setupLogging(self, taskDef, quantum):
 
         ButlerMDC.MDC("LABEL", label)
 
+        # Add the handler to the root logger.
+        # How does it get removed reliably?
```

Comment:
Do you mean if writeLogRecords is not called?

Reply:
Yes, maybe I should wrap the entire `execute` method in a `try/finally`.  On the other hand if I do fail to remove the handler eventually you'll get a `ResourceWarning` -- the tests did actually fail until I fixed some of the edge cases for calling writeLogRecords

Reply:
Or maybe make `setupLogging` to return a context manager and wrap all code in `with`?

Reply:
Hmm. Very good point.

Reply:
I also took the opportunity to add a context manager for the ButlerMDC setting so that it gets reset on exit.  Feel free to take a second look at this PR -- it's going to be a few hours before Jenkins completes.